File: console/console-api/src/routes/apiKeys.ts
import { Router, Request, Response } from 'express';
import { createHash, randomBytes } from 'crypto';
import pool from '../db';

const router = Router();

function generateKey(): string {
  return `ultr_${randomBytes(24).toString('base64url').substring(0, 32)}`;
}
function hashKey(key: string): string {
  return createHash('sha256').update(key).digest('hex');
}
function keyPrefix(key: string): string {
  return key.substring(0, 9);
}

router.get('/api-keys', async (req: Request, res: Response) => {
  try {
    const userId = req.headers['x-user-id'] as string;
    if (!userId) return res.status(401).json({ error: 'unauthorized' });
    const { rows } = await pool.query(
      'SELECT id, key_prefix, name, status, quota_limits, last_used_at, created_at FROM api_keys WHERE user_id = $1 ORDER BY created_at DESC', [userId]
    );
    res.json({ data: rows.map((k: any) => ({
      id: k.id, keyPrefix: k.key_prefix, name: k.name, status: k.status,
      quotaLimits: k.quota_limits, lastUsedAt: k.last_used_at, createdAt: k.created_at,
    }))});
  } catch (err) { res.status(500).json({ error: 'internal_error' }); }
});

router.post('/api-keys', async (req: Request, res: Response) => {
  try {
    const userId = req.headers['x-user-id'] as string;
    const orgId = req.headers['x-org-id'] as string;
    if (!userId || !orgId) return res.status(401).json({ error: 'unauthorized' });

    const { name, quotaLimits } = req.body;
    const plaintext = generateKey();
    const { rows: [key] } = await pool.query(
      `INSERT INTO api_keys (user_id, org_id, key_hash, key_prefix, name, quota_limits)
       VALUES ($1, $2, $3, $4, $5, $6) RETURNING id, key_prefix, name, created_at`,
      [userId, orgId, hashKey(plaintext), keyPrefix(plaintext), name, JSON.stringify(quotaLimits || { '*': 50000 })]
    );
    res.status(201).json({ data: { ...key, key: plaintext } });
  } catch (err) { res.status(500).json({ error: 'internal_error' }); }
});

router.delete('/api-keys/:id', async (req: Request, res: Response) => {
  try {
    const userId = req.headers['x-user-id'] as string;
    if (!userId) return res.status(401).json({ error: 'unauthorized' });
    await pool.query("UPDATE api_keys SET status = 'revoked', revoked_at = now() WHERE id = $1 AND user_id = $2", [req.params.id, userId]);
    res.json({ data: { status: 'revoked' } });
  } catch (err) { res.status(500).json({ error: 'internal_error' }); }
});

export default router;

File: auth-service/src/db/api_keys.rs
use sqlx::PgPool;
use uuid::Uuid;
use crate::types::ApiKey;
use crate::error::AppError;

pub async fn find_by_hash(pool: &PgPool, hash: &str) -> Result<Option<ApiKey>, AppError> {
    sqlx::query_as::<_, ApiKey>("SELECT * FROM api_keys WHERE key_hash = $1")
        .bind(hash)
        .fetch_optional(pool)
        .await
        .map_err(|e| AppError::Internal(format!("DB error: {}", e)))
}

pub async fn create(
    pool: &PgPool, user_id: &Uuid, org_id: &Uuid,
    key_hash: &str, key_prefix: &str, name: Option<&str>,
    quota_limits: &serde_json::Value,
) -> Result<ApiKey, AppError> {
    sqlx::query_as::<_, ApiKey>(
        "INSERT INTO api_keys (user_id, org_id, key_hash, key_prefix, name, quota_limits) \
         VALUES ($1, $2, $3, $4, $5, $6) RETURNING *"
    )
    .bind(user_id).bind(org_id).bind(key_hash).bind(key_prefix).bind(name).bind(quota_limits)
    .fetch_one(pool).await
    .map_err(|e| AppError::Internal(format!("DB error: {}", e)))
}

pub async fn revoke(pool: &PgPool, key_id: &Uuid) -> Result<(), AppError> {
    let rows = sqlx::query(
        "UPDATE api_keys SET status = 'revoked', revoked_at = now() WHERE id = $1 AND status = 'active'"
    )
    .bind(key_id).execute(pool).await
    .map_err(|e| AppError::Internal(format!("DB error: {}", e)))?;
    if rows.rows_affected() == 0 {
        return Err(AppError::ApiKeyNotFound);
    }
    Ok(())
}

pub async fn update_last_used(pool: &PgPool, key_id: &Uuid) -> Result<(), AppError> {
    sqlx::query("UPDATE api_keys SET last_used_at = now() WHERE id = $1")
        .bind(key_id).execute(pool).await
        .map_err(|e| AppError::Internal(format!("DB error: {}", e)))?;
    Ok(())
}

pub async fn list_by_user(pool: &PgPool, user_id: &Uuid) -> Result<Vec<ApiKey>, AppError> {
    sqlx::query_as::<_, ApiKey>(
        "SELECT * FROM api_keys WHERE user_id = $1 ORDER BY created_at DESC"
    )
    .bind(user_id).fetch_all(pool).await
    .map_err(|e| AppError::Internal(format!("DB error: {}", e)))
}
